spec/models/phone_number_authentication/authenticate_spec.rb

```diff
@@ -67,4 +67,86 @@
+      PhoneNumberAuthentication::Authenticate.call(otp: "xx12", password: "5489", phone_number: user.phone_number)
+      PhoneNumberAuthentication::Authenticate.call(otp: "xxxx", password: "5489", phone_number: user.phone_number)
+      PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "bad", phone_number: user.phone_number)
+      phone_number_authentication.reload
+      expect(phone_number_authentication.failed_attempts).to eq(4)
+      expect(phone_number_authentication.locked_at).to be_nil
+      time = nil
+      Timecop.freeze("January 1 2020 2:00 PM") do
+        time = Time.current
+        PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "bad", phone_number: user.phone_number)
+      end
+      phone_number_authentication.reload
+      expect(phone_number_authentication.failed_attempts).to eq(5)
+      expect(phone_number_authentication.locked_at).to eq(time)
+    end
+
+    it "returns locked out message if within locked window" do
+      user = FactoryBot.create(:user, password: "5489")
+      Timecop.freeze("January 1st 2020 12:00") do
+        user.phone_number_authentication.update(failed_attempts: 5, locked_at: Time.current)
+        Timecop.travel(2.minutes.from_now) do
+          result = PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "5489", phone_number: user.phone_number)
+          expect(result.success).to be false
+          expect(result.error_message).to eq("Your account is locked out for 18 minutes. Please wait and try again later.")
+        end
+        Timecop.travel(10.minutes.from_now) do
+          result = PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "5489", phone_number: user.phone_number)
+          expect(result.success).to be false
+          expect(result.error_message).to eq("Your account is locked out for 10 minutes. Please wait and try again later.")
+        end
+      end
+    end
+
+    it "unlocks the user after 20 minutes" do
+      user = FactoryBot.create(:user, password: "5489")
+      phone_number_authentication = user.phone_number_authentication
+      Timecop.freeze do
+        user.phone_number_authentication.update(failed_attempts: 5, locked_at: Time.current)
+        Timecop.travel(2.minutes.from_now) do
+          result = PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "5489", phone_number: user.phone_number)
+          expect(result.success).to be false
+          expect(result.error_message).to eq("Your account is locked out for 18 minutes. Please wait and try again later.")
+        end
+        Timecop.travel(21.minutes.from_now) do
+          phone_number_authentication.set_otp # handled by the mobile app I think?
+          phone_number_authentication.save
+          result = PhoneNumberAuthentication::Authenticate.call(otp: user.otp, password: "5489", phone_number: user.phone_number)
+          expect(result.success).to be true
```

Comment:
What do you think of `expect(result).to be_success`?

Reply:
i dig it